docker/Dockerfile.base

```diff
@@ -0,0 +1,251 @@
+		dst="$(echo "$src" | tr -d 3)"; \
+		[ -s "/usr/local/bin/$src" ]; \
+		[ ! -e "/usr/local/bin/$dst" ]; \
+		ln -svT "$src" "/usr/local/bin/$dst"; \
+	done \
+# ---------------------------------------------------------------------------------------------------------------------
+	# INSTALL PIP AND SETUPTOOLS
+    # https://github.com/docker-library/python/blob/f568f56f28fab0fe87b34db777e2c2861cef002b/3.11/slim-buster/Dockerfile#LL147C2-L166C15
+	&& wget -O get-pip.py "$PYTHON_GET_PIP_URL" \
+	&& echo "$PYTHON_GET_PIP_SHA256 *get-pip.py" | sha256sum -c - \
+	&& python get-pip.py \
+		--disable-pip-version-check \
+		--no-cache-dir \
+		--no-compile \
+		"pip==$PYTHON_PIP_VERSION" \
+		"setuptools==$PYTHON_SETUPTOOLS_VERSION" \
+	&& rm -f get-pip.py \
+	\
+	&& pip --version \
+# ---------------------------------------------------------------------------------------------------------------------
+    # INSTALLING SOUFFLE FROM SOURCE.
+    # Based on https://github.com/souffle-lang/souffle/blob/master/.github/images/oraclelinux-8/Dockerfile.
+    # However, some parts are different on Oracle Linux 9.
+    && mkdir /souffle-src \
+    && cd /souffle-src \
+    # Install build dependencies.
+    && dnf --enablerepo=ol9_codeready_builder install -y \
+        autoconf \
+        automake \
+        bash-completion \
+        bison \
+        clang \
+        cmake \
+        doxygen \
+        flex \
+        git \
+        libtool \
+        make \
+        pkg-config \
+        rpm-build \
+    && git clone --depth=1 https://github.com/souffle-lang/souffle/ -b 2.3 \
+    && cd souffle \
+    && cmake -S . -B build -DSOUFFLE_DOMAIN_64BIT=ON -DCMAKE_INSTALL_PREFIX="/usr/local" \
+    && nproc="$(nproc)" \
+    && cmake --build build --target install -j "$nproc" \
+    && cd / \
+    && rm -rf /souffle-src \
+    && souffle --version \
+# ---------------------------------------------------------------------------------------------------------------------
+    # CLEANING UP.
+    # We mark all unecessary packages to be removed while preserving the user installed packages.
+	&& dnf list installed | tail -n +2 | cut -d' ' -f1 | xargs -r dnf mark remove > /dev/null \
+    && [ -z "$USER_MANUAL_INSTALLED" ] || dnf mark install $USER_MANUAL_INSTALLED \
+    # Look for share libraries that are used by looking through the executables in /usr/local to preserve them.
+    # https://github.com/docker-library/python/blob/f568f56f28fab0fe87b34db777e2c2861cef002b/3.11/slim-buster/Dockerfile#LL111C2-L121C3
+    && find /usr/local -type f -executable -not \( -name '*tkinter*' \) -exec ldd '{}' ';' \
+		| awk '/=>/ { print $(NF-1) }' \
+		| sort -u \
+		| xargs -r dpkg-query --search \
+		| cut -d: -f1 \
+		| sort -u \
+		| xargs -r dnf mark install \
+    # We exclude systemd-udev because currently one of the build dependencies for Souffle pull this protected package.
+    # This command make sure that it won't break if systemd-udev is not pulled in the future.
```

Comment:
Why is it OK if in future `systemd-udev` is not pulled?

Reply:
What I meant was that if one of the build dependencies of Souffle decides to not depend on `systemd-udev` anymore, we won't mark `systemd-udev` as user installed. If a library is not installed but we mark it as `userinstalled`, then `dnf` will raise errors.